aten/src/ATen/native/mps/operations/Copy.mm

```diff
@@ -260,6 +260,25 @@ static bool copy_requires_temporaries(const Tensor& dst, const Tensor& src) {
   if (sourceBuffer == nil) return dst_;
   NSUInteger destOffset = dst.storage_offset() * dst.itemsize();
 
+  // In case of dtype change, first convert src "inplace"
+  if (src.dtype() != dst.dtype()) {
+    @autoreleasepool {
+      auto srcDType = getMPSDataType(src.scalar_type());
+      auto dstDType = getMPSDataType(dst.scalar_type());
+      auto srcShape = getMPSShape(src);
+      stream->commandBuffer();
+      MPSGraph* mpsGraph = make_mps_graph();
+      MPSGraphTensor* srcTensor = mps::mpsGraphRankedPlaceHolder(mpsGraph, src);
+      MPSGraphTensorData* srcData = [[[MPSGraphTensorData alloc]
+                                      initWithMTLBuffer:sourceBuffer shape:srcShape dataType:srcDType]
+                                     autorelease];
+      MPSGraphTensor *resultTensor = [mpsGraph castTensor:srcTensor toType:dstDType name:@"cast"];
```

Comment:
We should put this in a MPSGraphCache, so that we are not re-creating the Graph everytime. Please follow, the boiler-plate as is in done in [Activation functions](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/mps/operations/Activation.mm#L36). Which will be a single node graph, and should be very similar here.

Reply:
Thanks, this looks great @lhoenig